### DISCUSSION TITLE
CVC5 C API

### DISCUSSION THREAD

**Discussion by mister-walter** [cvc5/cvc5#9243] [Ideas]
At FMCAD @barrettcw brought up that a C API for CVC5 is coming, but just wanted to make a discussion to track this (let me know if an issue would be more appropriate).

My use case is doing interop between Common Lisp and SMT solvers. I developed some Common Lisp bindings for Z3's C API using [CFFI](https://cffi.common-lisp.dev/) and have been using them successfully, but it would be great to be able to use CVC5 in a similar manner!

Why is a C API helpful for me? CFFI, like many foreign function interface libraries, supports interoperating with C but doesn't support a lot of the complications that C++ has. These include templates, function name mangling (and the related issue of overloaded functions), C++ exceptions and methods. Libraries that are written and C++ and want to enable interop with C typically will provide C function wrappers around C++ functionality that deals with the aforementioned issues; this is [what Z3 does](https://github.com/Z3Prover/z3/blob/master/src/api/z3_api.h).

As a side note, there do exist libraries that allow for C++ interop in Common Lisp - for example, [SWIG](https://www.swig.org) (which supports other languages too), but it is not ideal for my application (in particular, it is not well integrated with Common Lisp's ASDF and requires the user to install it separately).

Let me know if I can help with this effort in any way!

**Comment by barrettcw** [comment]
We are hoping to begin work on a C API early next year.

In the meantime, we would be very interested in trying some of your sequence benchmarks.  It should be straightforward to have Z3 dump these in SMT-LIB format.

If you want to know more about our new sequence solver, check out this paper: https://theory.stanford.edu/~barrett/pubs/SNR+22-abstract.html

**Comment by aniemetz** [comment]
@mister-walter, just checking in that this is very high on our priority list and we will tackle this soon. Will keep you posted here.

**Reply by daniel-larraz** [reply]
@mister-walter , following up on this. The latest release of cvc5 (1.2.0) now includes a C API. You can find the documentation [here](https://cvc5.github.io/docs-ci/docs-main/api/c/c.html). I hope this helps.
